### ISSUE TITLE
gather_facts not skipped on skipped include

### ISSUE THREAD

**Issue by zigarn** [ansible/ansible#21528]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

ansible

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file =
  configured module search path = Default w/o overrides
```

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

N/A

##### SUMMARY

In Ansible 2.2.0.0, when using conditional on playbook include, none of the tasks of the included playbook where executed (not even skipped), the playbook was just not included with a message `skipping playbook include 'included.yml' due to conditional test failure`.

In ansible 2.2.1.0, the tasks are executed but marked as skipped. But the fact gathering is not!

For example, when the conditional is here to skip the playbook with fact gathering because the hosts are not accessible or thing like that, the previous behavior was perfect and the new one breaks it.

##### STEPS TO REPRODUCE

```ini
# hosts
localhost ansible_connection=local
```

```yaml
# main.yml
---
- include: specific.yml
  when: specific | default(True) | bool
```

```yaml
# specific.yml
---
- name: Specific play
  hosts: all
  tasks:
      - debug:
          msg: "Should only display on specific use"
```

```shell
ansible-playbook -i hosts main.yml
```

```shell
ansible-playbook -i hosts -e '{ "specific": false }' main.yml
```

##### EXPECTED RESULTS

```
$ ansible-playbook -i hosts -e '{ "specific": false }' main.yml

PLAY [Specific play] ***********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "Should only display on specific use"
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0

$
```

```
$ ansible-playbook -i hosts -e '{ "specific": false }' main.yml
skipping playbook include 'specific.yml' due to conditional test failure

PLAY RECAP *********************************************************************

$
```
##### ACTUAL RESULTS

```
$ ansible-playbook -i hosts main.yml

PLAY [Specific play] ***********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "Should only display on specific use"
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0

$
```

```
$ ansible-playbook -i hosts -e '{ "specific": false }' main.yml

PLAY [Specific play] ***********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
skipping: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0

$
```

**Comment by jctanner** [comment]
Changed title from "Conditionnal playbook behavior regression between 2.2.0.0 and 2.2.1.0" to "gather_facts not skipped on skipped include" because it took multiple reads before I realized this was about fact gathering and not the tasks.

**Comment by insequent** [comment]
^ I don't think that's accurate. The tasks themselves are running (or being skipped). It's causing our orchestration to break as well.

**Comment by jctanner** [comment]
The previous behavior was removed by https://github.com/ansible/ansible/commit/5b87951d6c99b515ceb07674cd862a2375328c7a#diff-dc700addce56e967a23db219cd13021b

Prior to that commit, failed conditionals on the "play level include" would cause the loading function to return a NoneType object instead of a playbook. Now, there will never be a NoneType and it will always be a playbook object. This then implies that https://github.com/ansible/ansible/blob/0b96d61162f996c7690ad20f0c2dc7d738525452/lib/ansible/playbook/__init__.py#L97 is dead code and should be removed because it will never be hit.

I confirmed with the author of 5b87951d6c99b515ceb07674cd862a2375328c7a that the current behavior is the desired behavior.

**Comment by insequent** [comment]
Ouch, then despite the fact that an entire playbook is being skipped, we now have to populate everything in the included playbook that might cause issues (i.e. any vars in ``when`` or ``with`` statements), since each task is technically being evaluated.

I can't say I agree with that intent, but we can work with it. Thanks for clarifying.

**Comment by zigarn** [comment]
I don't agree with this intent either. For the same reason as @insequent, we now need to be more defensive in the tasks of the included playbooks.
It's sad that all the tasks of the included -- but skipped -- playbook need to be parsed and evaluated, adding a lot of processing for dead tasks and a lot of output too.
Wouldn't have been easier to have a new type PlaybookSkipped? It would have avoid the NoneType problem and this current bug on fact gathering.

**Comment by jctanner** [comment]
@insequent @zigarn After a weekend of contemplation, I agree with you about the change. I created and merged #21734 as a stop gap to at least get us closer to the old behavior. We will probably make a better solution that might be similar to @zigarn's suggestion in the late 2.3/2.4 timeframe.

jimi-c is going to cherry pick the change from #21734 in the next 2.2 release

**Comment by insequent** [comment]
Awesome, thanks for looking into it @jctanner.

**Comment by zigarn** [comment]
Cool! Thanks for the fix @jctanner.

```shell
$ ansible --version
ansible 2.3.0 (devel b2bd75a408) last updated 2017/02/22 11:44:45 (GMT +200)
  config file =
  configured module search path = Default w/o overrides
```

```shell
$ ansible-playbook -i hosts main.yml

PLAY [Specific play] **********************************************************************

TASK [Gathering Facts] ********************************************************************
ok: [localhost]

TASK [debug] ******************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "Should only display on specific use"
}

PLAY RECAP ********************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```

```shell
$ ansible-playbook -i hosts -e '{ "specific": false }' main.yml

PLAY [Specific play] **********************************************************************

TASK [Gathering Facts] ********************************************************************
skipping: [localhost]

TASK [debug] ******************************************************************************
skipping: [localhost]

PLAY RECAP ********************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=0
```

**Comment by ovorobio** [comment]
@jctanner @jimi-c  is there any update re backporting this fix to 2.2 branch, please?

**Comment by mogthesprog** [comment]
I had raised this issue a few days ago #23173, it's root lies in the same issue, only in my case the play doesn't exist at all, so it errors out. I'll close the other issue in favour of this.

Backporting this fix to 2.2 would be much appreciated :)

**Comment by RalfJung** [comment]
I am still having this issue with Ansible 2.5.2: When an imported playbook is skipped because of a tag, it still has its facts gathered. I have plenty of entries like this in my site.yml:
```
- hosts: bind
  roles:
  - bind
  tags: bind
```
and then I use tags to deploy only specific roles, but ansible still spends tons of time just gathering facts for all the roles that are not executed.

From what I understand, #21734 fixed this for conditional includes, but the issue remains for tags?

**Comment by bcoca** [comment]
this is fixed in current (2.8) version of Ansible, now it correctly skips the gathering task.
